File: descripe/imapp_modisl2/src/leocat/src/MOD_PRAlg29/snow_mask.c
/*
!C ********************************************************************
!Description:

  Function snow_mask

  Subroutine which implements a quick and dirty version of the snow
  algorithm by Dorothy Hall, George Riggs and Vince Salmonson.

!Input arguments:
  none

  Input values stored in variables found in pixel.h, granule.h, 
  mask_processing _constants.h, and thresholds.h.

!Output arguments:
  none

  Output stored in 'pxin.ndsi_snow' found in pixel.h.

!Revision History:
  R. Frey           10/2007  Converted to C

!Team-unique Header:

!References and Credits:

!END ******************************************************************/

/* Includes */

#include <stdio.h>
#include <math.h>
#include "granule.h"
#include "pixel.h"
#include "mask_processing_constants.h"
#include "thresholds.h"

void snow_mask()

{ 

// Declarations
   int I_bad;
   
   float m04;
   float m02;
   float m26;
   float m20;
   float m29;
   float m31;
   float m35;
   float nirband;
   float ndsi;
   float diff;
   float diff2;
   float sth85_11;
   float sth37_11;

// Initializations
   m04 = pxin.rad[3];
   m02 = pxin.rad[1];
   m26 = pxin.rad[25];
   m20 = pxin.rad[19];
   m29 = pxin.rad[28];
   m31 = pxin.rad[30];
   m35 = pxin.rad[34];

   diff = 0.0;
   diff2 = 0.0;
   
// Select NIR channel for NDSI based on platform name.
// Band 6 for Terra, band 7 for Aqua.
   if(sid_terra) {
     nirband = pxin.rad[5];
   }     
   else if(sid_aqua) {
     nirband = pxin.rad[6];
   }  
   else {
     printf("Platform name not recognized in snow_mask \n");
   }

   I_bad = rint(bad_data);
   
// Begin snow/ice detection.
   
// First, check the 11 micron brightness temperature.
   if (rint(m31) != I_bad && m31 <= sm_bt11[0]) {

//   Get NDSI index
     if (rint(m04) != I_bad && rint(nirband) != I_bad) {
       ndsi  = (m04 - nirband) / (m04 + nirband);
//     printf("NDSI value is: %f %f %f %f %f %f \n", ndsi, sm_ndsi[0], m04, nirband, m02, sm_ref2[0]);

       if (ndsi > sm_ndsi[0] && m02 > sm_ref2[0]) { 
         pxin.ndsi_snow = 1;

//       Check for false snow detection.
//       Make sure NDSI is not flagging a thin cirrus cloud.
         if ( !pxin.Greenland && pxin.land) {
           if (rint(m26) != I_bad && rint(m35) != I_bad) {
             if (m26 > sm_ref3[0] && m35 < sm_co2[0]) {
               pxin.ndsi_snow = 0;
             }
           }
         }
         
//       If in sunglint region, disregard if between -60 and 50 degrees latitude.
         if(pxin.water && pxin.sunglint && pxin.lat <= 50.0 && pxin.lat >= -60.0) {
           pxin.ndsi_snow = 0;
         }
         
//       Check for ice clouds mis-identified as snow. Note modified BTD thresholds for Greenland.
         if(rint(m29) != I_bad && rint(m31) != I_bad) {
           diff = m29 - m31;
           sth85_11 = sm85_11[0];
           if(rint(m20) != I_bad && !(pxin.Greenland && pxin.land)) {
             diff2 = m20 - m31;
             sth37_11 = sm37_11[0];
             if(diff >= sth85_11 && diff2 >= sth37_11) {
               pxin.ndsi_snow = 0;
             }
           }  
           else {
             if(pxin.Greenland && pxin.land) {
               sth85_11 = sm85_11[0] + 1.5;
             }  
             else {
               sth85_11 = sm85_11[0];
             }
             if(diff >= sth85_11) {
               pxin.ndsi_snow = 0;
             }
           }
//         printf("ndsi check ice: %d %d %f %f %f %f %f \n", pxin.Greenland, pxin.land, diff, sth85_11, m20, diff2, sth37_11);
         }
           
//       Check for water clouds mis-identified as snow.
         if(m20 != I_bad && m31 != I_bad) {
           if( !(pxin.land && pxin.Greenland) ) {
             diff = m20 - m31;
             if(pxin.hi_elev) {
               sth37_11 = sm37_11hel[0];
             }  
             else {
               sth37_11 = sm37_11[0];
             }
             if(diff >= sth37_11) {
               pxin.ndsi_snow = 0;
             }
           }
//         printf("ndsi check water: %f %f \n", diff, sth37_11);
         }

//       Check value of NIR reflectance.         
         if( !pxin.hi_elev && !(pxin.Greenland && pxin.land) ) {
           if(nirband > sm_mnir[0]) {
             pxin.ndsi_snow = 0;
           }
//         printf("ndsi check NIR refl: %f %f \n", nirband, sm_mnir[0]);
         }
        

       } 
     }
   }
    

}
   

File: descripe/imapp_modisl2/src/cloudmask/src/snow_mask.f
      subroutine snow_mask(pxldat,plat,land,snglnt,water,hi_elev,
     +                     Greenland,ndsi_snow)

      implicit none
      save

c ... Common statement for debug purposes
      common / bug / debug, h_output

c---------------------------------------------------------------------
c!F77
c
c!Description:
c       Subroutine which implements a quick and dirty version of
c       the snow algorithm by Dorothy Hall, George Riggs and 
c       Vince Salmonson.
c
c!Input Parameters:
c pxldat        Array containing reflectance or brightness temperatures
c               for all bands for a single pixel
c plat          Latitude of current pixel
c land          Flag indicating a pixel to be processed as land
c snglnt        Logical variable flagging background as sunglint 
c               contaminated
c water         Logical variable indicating water surface for current 
c               pixel
c hi_elev       Logical flag indicating elevation > 2000 meters
c Greenland     Logical flag indicating location is "Greenland"
c
c!Output Parameters:
c ndsi-snow     logical variable indicating the prescence of snow in a
c               FOV according to the NDSI test
c
c!Revision History:
c 10/04  Collection 5b  R. Frey
c Removed definition of 'Greenland'.
c Removed tests on 1.38 um, 'sm_mnir', 3.7-11 um BTD in Greenland.
c Added 1.5K to 8.5-11 um BTD test threshold in Greenland.
c
c!Team-unique Header:
c
c!References and Credits:
c See snow product ATBD.
c
c!END
c---------------------------------------------------------------------

      include 'global.inc'
      include 'snow_mask.inc'
      include 'platform_name.inc'
c
c ... scalar arguments ..
      logical snglnt,water,ndsi_snow,hi_elev,land,Greenland
      real plat
c ...
c ... array arguments ..
      real pxldat(inband)
c ...
c ... local scalars ..
      real ndsi,masv88,masv55,masv188,masir13,masir11,
     +     masir85,masir37,diff,sth37_11,masnir,diff2,sth85_11
      integer h_output,debug,I_bad
c
c ... identification 
      masv55 = pxldat(4)
      masv88 = pxldat(2) 
      masv188 = pxldat(26)
      masir37 = pxldat(20)
      masir85 = pxldat(29)
      masir11 = pxldat(31)
      masir13 = pxldat(35)
c     Select NIR channel for NDSI based on platform name.
c     Band 6 for Terra, band 7 for Aqua.
      if(platform_name .eq. 'Terra') then
        masnir = pxldat(6)
      else if(platform_name .eq. 'Aqua ') then
        masnir = pxldat(7)
      else
        call message( 'snow_mask', 'Platform name not recognized' //
     &    ' [OPERATOR ACTION: Contact SDST]', 0, 2)
      endif

c ... debug statement ............................................
      if (debug .gt. 0) then
        write(h_output,'(10x/,''Subroutine Snow_Mask '',/)')
      endif
c ................................................................

c ... Initialize
      ndsi = 0.0
      ndsi_snow = .false.
      I_bad = nint(bad_data)

c ................................................................

c ... First, check the 11 micron brightness temperature.
      if (nint(masir11) .ne. I_bad .and. masir11 .le. sm_bt11(1)) then

c ...    perform the NDSI on the current pixel
         if (nint(masv55) .ne. I_bad .and. nint(masnir).ne.I_bad) then
           ndsi  = (masv55 - masnir) / (masv55 + masnir)

           if ( (ndsi .gt. sm_ndsi(1)) .and. (masv88 .gt. sm_ref2(1)) ) 
     +     then

             ndsi_snow = .true.

c            Check for false snow detection.

c            Now, make sure NDSI is not flagging a thin cirrus
             if( .not. (Greenland .and. land) ) then
               if(nint(masv188).ne.I_bad.and.nint(masir13).ne.I_bad) then
                 if (masv188 .gt. sm_ref3(1) .and. masir13 .lt. sm_co2(1)) then
                   ndsi_snow = .false.
                 endif
               endif
             endif
c            If in sunglint region, disregard if between -60 and 50 lat.
             if(water .and. snglnt .and. plat .le. 50.0 .and.
     +          plat .ge. -60.0) then
               ndsi_snow = .false.
             end if
c            Check for ice clouds mis-identified as snow. Note modified
c            BTD thresholds for Greenland.
             if(masir85 .ne. I_bad .and. masir11 .ne. I_bad) then
               diff = masir85 - masir11
               sth85_11 = sm85_11(1)
               if(masir37 .ne. I_bad .and. (.not. (Greenland .and. land)) ) then
                 diff2 = masir37 - masir11
                 sth37_11 = sm37_11(1)
                 if(diff .ge. sth85_11 .and. diff2 .ge. sth37_11) then
                   ndsi_snow = .false.
                 end if
               else
                 if(Greenland .and. land) then
                   sth85_11 = sm85_11(1) + 1.5
                 else
                   sth85_11 = sm85_11(1)
                 end if
                 if(diff .ge. sth85_11) then
                   ndsi_snow = .false.
                 end if
               end if
             end if
c            Check for water clouds mis-identified as snow.
             if(masir37 .ne. I_bad .and. masir11 .ne. I_bad) then
               if( .not. (land .and. Greenland) ) then
                 diff = masir37 - masir11
                 if(hi_elev) then
                   sth37_11 = sm37_11hel(1)
                 else
                   sth37_11 = sm37_11(1)
                 end if
                 if(diff .ge. sth37_11) then
                   ndsi_snow = .false.
                 end if
               end if
             end if

             if( (.not. hi_elev) .and. (.not. (Greenland .and. land)) ) then
               if(masnir .gt. sm_mnir(1)) then
                 ndsi_snow = .false.
               end if
             end if

           endif 
         endif
      endif

c ... debug statement ............................................
      if (debug .gt. 0) then
        write(h_output,'(15x,'' snow mask variables '',L4)') ndsi_snow
        write(h_output,'(2x,''masv55 masnir ndsi masv188 masv88 masir13
     +      snglnt hi_elev ndsi_snow '',/,6f8.4,f7.2,3L8)') masv55,masnir,ndsi,
     +                   masv188,masv88,sth85_11,masir13,snglnt,hi_elev,ndsi_snow
        write(h_output,'(''IR vals '',3f10.2)') masir85,masir11,masir37
      endif
c ................................................................

      return
      end
